include/deal.II/lac/petsc_ts.h

```diff
@@ -0,0 +1,706 @@
+        comm = pcomm;
+      return comm;
+    }
+
+    /**
+     * Reset the subobjects associated with TS, do not change customization.
+     */
+    void
+    reinit()
+    {
+      AssertTS(TSReset(ts));
+    }
+
+    /**
+     * Reset the subobjects associated with TS, changes customization according
+     * to data.
+     */
+    void
+    reinit(const TimeStepperData &data)
+    {
+      reinit();
+
+      // Solver type
+      if (data.tstype.size())
+        AssertTS(TSSetType(ts, data.tstype.c_str()));
+
+      // Time and steps limits
+      AssertTS(TSSetTime(ts, data.initial_time));
+      if (data.final_time > data.initial_time)
+        AssertTS(TSSetMaxTime(ts, data.final_time));
+      if (data.initial_step_size > 0.0)
+        AssertTS(TSSetTimeStep(ts, data.initial_step_size));
+      if (data.max_steps >= 0)
+        AssertTS(TSSetMaxSteps(ts, data.max_steps));
+      // Decide how to end the integration. Either stepover the final time or
+      // match it.
+      AssertTS(TSSetExactFinalTime(ts,
+                                   data.match_step ?
+                                     TS_EXACTFINALTIME_MATCHSTEP :
+                                     TS_EXACTFINALTIME_STEPOVER));
+
+      // Adaptive tolerances
+      PetscReal atol = data.absolute_tolerance > 0.0 ?
+                         data.absolute_tolerance :
+                         static_cast<PetscReal>(PETSC_DEFAULT);
+      PetscReal rtol = data.relative_tolerance > 0.0 ?
+                         data.relative_tolerance :
+                         static_cast<PetscReal>(PETSC_DEFAULT);
+      AssertTS(TSSetTolerances(ts, atol, nullptr, rtol, nullptr));
+
+      // Adaptive time stepping
+      TSAdapt tsadapt;
+      AssertTS(TSGetAdapt(ts, &tsadapt));
+
+      // Use adaptive time stepping only if explicitly requested
+      if (data.absolute_tolerance == 0.0 && data.relative_tolerance == 0.0)
+        {
+          AssertTS(TSAdaptSetType(tsadapt, TSADAPTNONE));
+        }
+      else
+        {
+          AssertTS(TSAdaptSetType(tsadapt, TSADAPTBASIC));
+        }
+
+      // Step limits
+      PetscReal hmin = data.minimum_step_size > 0 ?
+                         data.minimum_step_size :
+                         static_cast<PetscInt>(PETSC_DEFAULT);
+      PetscReal hmax = data.maximum_step_size > 0 ?
+                         data.maximum_step_size :
+                         static_cast<PetscInt>(PETSC_DEFAULT);
+      AssertTS(TSAdaptSetStepLimits(tsadapt, hmin, hmax));
+    }
+
+
+    /**
+     * Pass the preconditioning matrix only.
+     * When used with 'setup_jacobian' and 'solve_for_jacobian_system', PETSc
+     * will approximate the linear system matrix-vector product using an
+     * internal matrix-free representation.
+     * When used with 'implicit_jacobian' or 'explicit_jacobian',
+     * PETSc will use the same matrix for both preconditioning and
+     * matrix-vector products.
+     */
+    void
+    reinit_matrices(PMatrixType &P)
```

Comment:
The default for the template argument is `PMatrixType == AMatrixType`, which allows people to accidentally call `reinit_matrices(A)`. My preference is to call this function `reinit_preconditioner_matrix()` to make it clear what matrix it corresponds to.

Separately, though, I don't understand the documentation. What is it that PETSc does when you only provide a preconditioner, but not a system matrix? Isn't the latter always necessary?

Reply:
I have added some text in the docs to clarify this point and renamed it in `reinit_matrix(A)`.